### PR TITLE
adding flag to control number of uplink ports when creating a new distributed switch

### PR DESCRIPTION
**Pull request by haydonryan**
## Description

When creating distributed and nested switches, you will often want to provide only a single uplink port, or in the case of a 4 port nic, utilize 4 ports.  This feature was missing from dvs.create.

I also added a local check to ensure that you don't try and create a DVS with 0 or invalid ports, as vSphere won't let you do that.

No dependencies

Closes: #2314

## Type of change

Please mark options that are relevant:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
- I used the code that  @dougm provided in the issue, and put modified the code. I compared what I received in the simulator with the output Doug provided by running the Powershell command.

- I then powered up my homelab and tested this against VSphere 7.0 u2 and created a DVS with 1 uplink port, 3, 1030. 
Added a default of 2 ports, and added the CLI flags to override.  Tested again with no flag, 1,3,1030.

**Test Configuration**:
* Toolchain: go version go1.14.2 linux/amd64

## Checklist:

- [ ] My code follows the CONTRIBUTION [guidelines](./../CONTRIBUTING.md) of
  this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works  - I can't find where to add tests to this?
- [ ] New and existing unit tests pass locally with my changes - I can't find where to add tests to this? ran go test but didnt' see any tests (other than main) run.
- [x] Any dependent changes have been merged

### CODE REVIEW

File: govc/dvs/create.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ type create struct {
 	configSpec *types.VMwareDVSConfigSpec
 
 	dProtocol string
+
+	numUplinkPorts int32
```
Comment (dougm) [inline govc/dvs/create.go:38]: Can use `int` here instead of `int32`, to avoid the later conversion?

File: govc/dvs/create.go [hunk 2]
```diff
@@ -49,9 +51,12 @@ func (cmd *create) Register(ctx context.Context, f *flag.FlagSet) {
 	cmd.DVSCreateSpec.ConfigSpec = cmd.configSpec
 	cmd.DVSCreateSpec.ProductInfo = new(types.DistributedVirtualSwitchProductSpec)
 
+	cmd.numUplinkPorts = 2 // Set Default Uplink ports to 2
```
Comment (dougm) [inline govc/dvs/create.go:54]: I think the default should be '0' to maintain the existing behavior, what do you think?

File: govc/dvs/create.go [hunk 3]
```diff
@@ -49,9 +51,12 @@ func (cmd *create) Register(ctx context.Context, f *flag.FlagSet) {
...
+
 	f.StringVar(&cmd.ProductInfo.Version, "product-version", "", "DVS product version")
 	f.Var(flags.NewInt32(&cmd.configSpec.MaxMtu), "mtu", "DVS Max MTU")
 	f.StringVar(&cmd.dProtocol, "discovery-protocol", "", "Link Discovery Protocol")
+	f.Var(flags.NewInt32(&cmd.numUplinkPorts), "num-uplinks", "Number of Uplinks")
```
Comment (dougm) [inline govc/dvs/create.go:59]: Looks like you can just use `f.IntVar(&cmd.numUplinkPorts, ...)` here.  We only need to use `f.Var` + `flags.NewX` for spec fields sent over the wire.  In this case `cmd.numUplinkPorts` is just a local counter.

File: govc/dvs/create.go [hunk 4]
```diff
@@ -94,6 +99,18 @@ func (cmd *create) Run(ctx context.Context, f *flag.FlagSet) error {
...
+	numUplinkPorts := int(cmd.numUplinkPorts)
+
+	if numUplinkPorts > 0 {
+		var policy types.DVSNameArrayUplinkPortPolicy
+		for i := 0; i < numUplinkPorts; i++ {
+			policy.UplinkPortName = append(policy.UplinkPortName, fmt.Sprintf("Uplink %d", i+1))
+		}
+		cmd.configSpec.UplinkPortPolicy = &policy
+	} else {
+		return errors.New("Must specify more than one uplink port")
```
Comment (dougm) [inline govc/dvs/create.go:111]: This isn't required to create a DVS though, is it?  Seems like we can keep the default as `0`, without erroring here.

**Review by dougm — COMMENTED** [review]
Thanks @haydonryan !  No need to add attribution for my part of the patch, but there are some docs on how-to here: https://docs.github.com/en/github/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

One other request, if you could change the commit message to include the 'govc: ' prefix, that'll help the CHANGELOG generator: https://github.com/vmware/govmomi/blob/master/CONTRIBUTING.md#format-of-the-commit-message

See CHANGELOG.md for other examples:
`govc: add dvs.create '-num-uplinks' flag`

File: govc/dvs/create.go (hunk 2 above)
Comment (haydonryan) [reply govc/dvs/create.go:54]: My only concern around that is someone reading the code will see cmd.numUplinkPorts = 0 and assume that this creates zero ports when it actually skips the if block and creates 2 uplinks.

File: govc/dvs/create.go (hunk 1 above)
Comment (haydonryan) [reply govc/dvs/create.go:38]: Sounds Good, but how about we use uint since the number should never be negative?

File: govc/dvs/create.go (hunk 4 above)
Comment (haydonryan) [reply govc/dvs/create.go:111]: I was trying to have this cover if someone put in a negative number or hardcoded zero.  Originally I wasn't sure if there was an upper limit on the number of uplinks, but there doesn't appear to be.  Happy to remove if not appropriate.

File: govc/dvs/create.go (hunk 1 above)
Comment (dougm) [reply govc/dvs/create.go:38]: Sure, you can do that and use `flag.Uint()`

File: govc/dvs/create.go (hunk 2 above)
Comment (dougm) [reply govc/dvs/create.go:54]: I think it's ok to leave it as: `f.UintVar(&cmd.numUplinkPorts, "num-uplinks", 0, "Number of Uplinks")`
And let the `CreateDVS` method do what it does by default.

File: govc/dvs/create.go (hunk 4 above)
Comment (dougm) [reply govc/dvs/create.go:111]: Looks like if you use `f.UintVar(...)`, then the Go flag package will complain about a negative value, and `0` value will just result in vSphere's default behavior as the command does currently.

**Pull request closed without merging**